src/lib.rs

```diff
@@ -1,6 +1,98 @@
+//!
+//! While for the most part an IOMMU may have device level granularity, any system is susceptible
+//! to reduced granularity. The IOMMU API therefore supports a notion of IOMMU groups. A group is
+//! a set of devices which is isolatable from all other devices in the system. Groups are therefore
+//! the unit of ownership used by VFIO.
+//!
+//! While the group is the minimum granularity that must be used to ensure secure user access, it's
+//! not necessarily the preferred granularity. In IOMMUs which make use of page tables, it may be
+//! possible to share a set of page tables between different groups, reducing the overhead both to
+//! the platform (reduced TLB thrashing, reduced duplicate page tables), and to the user
+//! (programming only a single set of translations). For this reason, VFIO makes use of a container
+//! class, which may hold one or more groups. A container is created by simply opening the
+//! /dev/vfio/vfio character device.
+//!
+//! This crate is a safe wrapper around the Linux kernel's VFIO interfaces, which offering safe
+//! wrappers for:
+//! - [VFIO Container](struct.VfioContainer.html) using the `VfioContainer` structure
+//! - [VFIO Device](struct.VfioDevice.html) using the `VfioDevice` structure
+//!
+//! # Platform support
+//!
+//! - x86_64
+//!
+//! **NOTE:** The list of available ioctls is not extensive.
+
 #![deny(missing_docs)]
-//! Dummy crate needs high-level documentation.
-/// Dummy public function needs documentation.
-pub fn it_works() {
-    assert!(true);
+
+#[macro_use]
+extern crate vmm_sys_util;
+
+mod vfio_device;
+mod vfio_ioctls;
+
+pub use vfio_device::{VfioContainer, VfioDevice, VfioError, VfioIrq};
+
+use std::mem::size_of;
+
+/// Returns a `Vec<T>` with a size in bytes at least as large as `size_in_bytes`.
+fn vec_with_size_in_bytes<T: Default>(size_in_bytes: usize) -> Vec<T> {
+    let rounded_size = (size_in_bytes + size_of::<T>() - 1) / size_of::<T>();
+    let mut v = Vec::with_capacity(rounded_size);
+    for _ in 0..rounded_size {
+        v.push(T::default())
+    }
+    v
+}
+
+/// The kvm API has many structs that resemble the following `Foo` structure:
+///
+/// ```
+/// #[repr(C)]
+/// pub struct __IncompleteArrayField<T>(::std::marker::PhantomData<T>);
+/// #[repr(C)]
+/// struct Foo {
+///    some_data: u32,
+///    entries: __IncompleteArrayField<u32>,
+/// }
+/// ```
+///
+/// In order to allocate such a structure, `size_of::<Foo>()` would be too small because it would not
+/// include any space for `entries`. To make the allocation large enough while still being aligned
+/// for `Foo`, a `Vec<Foo>` is created. Only the first element of `Vec<Foo>` would actually be used
+/// as a `Foo`. The remaining memory in the `Vec<Foo>` is for `entries`, which must be contiguous
+/// with `Foo`. This function is used to make the `Vec<Foo>` with enough space for `count` entries.
+pub fn vec_with_array_field<T: Default, F>(count: usize) -> Vec<T> {
```

Comment:
The FAM helper from vmm_sys_util can't be used here because there data struct layout doesn't satisfy FAM's expectation.

Reply:
I looked again at the FamStructWrapper in vmm-sys-util and it looks like it is too restrictive. The structure `vfio_irq_set` is FAM because it has an incomplete array and the number of elements in that array. The same applies to the wrapper `vfio_region_info_with_cap`.

I think this is a miss on our part. Can you open an issue in vmm-sys-util to make FamStructWrapper less restrictive so that we can use it here as well? Than, we can also add an issue in this repository.